### ISSUE TITLE
Forward and reverse mode differentiation at the same time

### ISSUE THREAD

**Issue by h-spiess** [jax-ml/jax#1868]
Hey,

is it possible to do forward mode and reverse mode differentiation at the same time in JAX for different sets of parameters, e.g., hyperparameter and network weights, respectively?

Example:

Neural Network with a scalar hyperparameter, e.g., the learning rate. 
The goal would be then to train the network using standard reverse mode differentiation, while simultaneously accumulating the gradient of the scalar parameter in forward mode (reverse-mode would require to store all intermediate weight states (= similar to backprop through time)).

Unfortunately i haven't found a way to do this in JAX yet.

Thanks a lot for your help.

Heiner

**Comment by hawkinsp** [comment]
Can you give a small, self-contained, concrete example of what you want to compute? I'm a little hazy from your description, but if you made the example concrete then I think we might be able to make progress.

**Comment by h-spiess** [comment]
Thanks, i will try to present a simple example to see if what i want to do is possible at all before going to the more complicated problem described before.

```
import jax.numpy as np
from jax import random

key = random.PRNGKey(0)

key, W_key, b_key = random.split(key, 3)
W = random.normal(W_key, (3,))
b = random.normal(b_key, ())

a = 2.0

def predict(W, b, a, inputs):
    return sigmoid(np.dot(inputs, W - a) + b)

# Build a toy dataset.
inputs = np.array([[0.52, 1.12,  0.77],
                   [0.88, -1.08, 0.15],
                   [0.52, 0.06, -1.30],
                   [0.74, -2.49, 1.39]])
targets = np.array([True, True, False, True])

# Training loss is the negative log-likelihood of the training examples.
def loss(W, b, a):
    preds = predict(W, b, a, inputs)
    label_probs = preds * targets + (1 - preds) * (1 - targets)
    return -np.sum(np.log(label_probs))

# TODO
# calculate gradient w.r.t. 'a' by forward-mode
# at the same time: calculate gradients of 'W' and 'b' by reverse mode

# why: e.g. want to update 'a' after a whole epoch of updating 'W' and 'b'
```

**Comment by mattjj** [comment]
Sorry for neglecting to answer this for like 5.5 years. Hope you haven't been stuck waiting for us this whole time.

Here's the ~one-liner you want:

```python
loss_value, a_deriv, (W_grad, b_grad) = \
    jax.jvp(lambda a: jax.value_and_grad(loss, (0, 1))(W, b, a), [a], [1.0], has_aux=True)
```

The key bits are
* use `jax.jvp` for forward-mode on `a`
* use `jax.grad(loss, argnums=(0, 1))` for reverse-mode on `W` and `b`
* use `has_aux=True` on the `jax.jvp` call so that we don't pay attention to how changing `a` will perturb the values of `W_grad` and `b_grad`, and instead just look at its effect on the loss
